src/rubyLsp.ts

```diff
@@ -0,0 +1,401 @@
+import path from "path";
+
+import * as vscode from "vscode";
+import { Range } from "vscode-languageclient/node";
+
+import { Telemetry } from "./telemetry";
+import DocumentProvider from "./documentProvider";
+import { Workspace } from "./workspace";
+import { Command, STATUS_EMITTER, pathExists } from "./common";
+import { VersionManager } from "./ruby";
+import { StatusItems } from "./status";
+import { TestController } from "./testController";
+import { Debugger } from "./debugger";
+
+// The RubyLsp class represents an instance of the entire extension. This should only be instantiated once at the
+// activation event. This class controls all of the existing workspaces, telemetry and handles all commands
+export class RubyLsp {
+  private readonly workspaces: Map<string, Workspace> = new Map();
+  private readonly telemetry: Telemetry;
+  private readonly context: vscode.ExtensionContext;
+  private readonly statusItems: StatusItems;
+  private readonly testController: TestController;
+  private readonly debug: Debugger;
+
+  constructor(context: vscode.ExtensionContext) {
+    this.context = context;
+    this.telemetry = new Telemetry(context);
+    this.testController = new TestController(
+      context,
+      this.telemetry,
+      this.currentActiveWorkspace.bind(this),
+    );
+    this.debug = new Debugger(context, this.currentActiveWorkspace.bind(this));
+    this.registerCommands(context);
+
+    this.statusItems = new StatusItems();
```

Comment:
It looks like we'll call `extension.activate()` right after this constructor is executed. Is it possible to initialise `StatusItems` there with workspace instead?

Reply:
If VS Code is opened with multiple workspaces and no files are currently opened, then the current active workspace is `undefined`. That would require handling the `undefined` scenario in the status items, which is why I ended up with this approach.

Basically, the status items are first instantiated with backup values. The moment a file is opened or switched, the active workspace changes and then we immediately update to show the information relevant to that workspace.